### ISSUE TITLE
[Question] Model uncertainty in the presence of missing data

### ISSUE THREAD

**Issue by LMikeH** [pytorch/botorch#170]
# 🚀 Feature Request

Hello,

This looks like a really cool BO project. I'm really glad a BO project is being built on Pytorch. I need some clarification around modeling uncertainty with GPytorch.

## Motivation
I have had difficulty in getting models from GPytorch to accurately predict epistemic uncertainty. Generally I see 1D regression tutorials with evenly spaced toy data sets that don't give the opportunity to show whether or not the model's uncertainty grows in regions where data is lacking like I'd expect. I noticed in your example this is also the case. Is your wrapped GPytorch model using advanced settings to ensure an accurate and expressive posterior that doesn't underestimate epistemic uncertainty?

## Pitch
It would be very helpful to have some discussion on expressive posteriors, especially if the model has limitations for data sets that has regions of missing data. If the model is capable of accurate epistemic uncertainty predictions it would be helpful to use an example that shows this more clearly. Ultimately, I think that models being able to tell me what they don't know is one the most important things for BO.

Thanks,
Michael Huang

**Comment by Balandat** [comment]
So this is a pretty broad question, let me answer it in a general and more specific way.

Generally, BoTorch's design philosophy is such that you should be able to use pretty much any probabilistic model that suits your particular need. Our goal is to make it easy to use BO in a modular way, so you have a lot of freedom to play around with the model. The BoTorch Model APIs are lightweight and set up in a way so this should be relatively straightforward to do. This kind of agency means that to some extent we are outsourcing the understanding of some of the modeling to the user. 

More specifically, GP models are typically quite good at providing well-calibrated uncertainty estimates if there are regions of missing data. The pre-packaged GP models in `botorch.models` are, as you observed, mostly wrappers around basic gpytorch models with specific choices for the mean and covariance functions, and the priors of the respective hyperparameters. These are pretty decent all-round models and should do a reasonable job with missing data (as long as you normalize and standardize your data). For instance, if I remove a data points from the center of the interval used in [batch_mode_cross_validation tutorial](https://botorch.org/tutorials/batch_mode_cross_validation), you'll end up with a posterior that looks quite reasonable (see [this notebook](https://github.com/pytorch/botorch/files/3272923/batch_mode_cross_validation_missing.ipynb.txt) for the full code):

![missing_data](https://user-images.githubusercontent.com/1605878/59210730-19a86400-8b63-11e9-9c94-fefa78caf1ee.png)

> I have had difficulty in getting models from GPytorch to accurately predict epistemic uncertainty. 

In what sense? If this is purely about modeling in gpytorch, I'd recommend raising a question on the gpytorch github.


I hope this helps.

**Comment by LMikeH** [comment]
Hi Balandat,

That clears some things up. I think I just need to do some digging to understand how you built your GPytorch models and selection of priors works. When I had previously used ExactGP model with RBF kernel pretty much straight from the tutorial it seemed to struggle with some of the data I was trying to fit. Maybe it was just due to not normalizing my data properly. 

Thanks for clarifying,
Michael

**Comment by Balandat** [comment]
> That clears some things up. I think I just need to do some digging to understand how you built your GPytorch models and selection of priors works. 

Great. The SingleTaskGP is a relatively vanilla `ExactGP`, with the kernel, likelihood and respective priors selected [here](https://github.com/pytorch/botorch/blob/master/botorch/models/gp_regression.py#L72-L103).

> Maybe it was just due to not normalizing my data properly.

Especially when using priors (as the botorch models do), this is very important. Make sure to have your input `X` normalized to `[0, 1]^d` and your observations standardized to zero mean and unit variance.

**Comment by Balandat** [comment]
closing this out for now, feel free to re-open if necessary.

**Comment by eytan** [comment]
Hi Michael,

I would suggest making to scale your Xs to be normalized between 0 and 1 and standardize your Ys, otherwise both GPyTorch and the priors we use in won’t work well.
